app/ydoc-shared/src/ast/tree.ts

```diff
@@ -102,6 +106,26 @@ export abstract class Ast {
     return id
   }
 
+  /** Returns whether this type can be an expression. */
+  isAllowedInExpressionContext(): boolean {
+    return false
+  }
+
+  /** Returns whether this type can be a statement. */
+  isAllowedInStatementContext(): boolean {
+    return false
+  }
+
+  /** Returns whether this type can be an expression. */
+  isExpression(): this is Expression {
+    return this.isAllowedInExpressionContext()
+  }
+
+  /** Returns whether this type can be a statement. */
+  isStatement(): this is Statement {
+    return this.isAllowedInStatementContext()
+  }
+
```

Comment:
Duplicated docs. The fact one pair of functions uses another does not help.

+ good occasion to fill docs placeholder for this class (helpful as a root class of a large structure)